ml-agents/mlagents/trainers/env_manager.py

```diff
@@ -109,6 +118,18 @@ def advance(self):
         return num_step_infos
 
     def _process_step_infos(self, step_infos: List[EnvironmentStep]) -> int:
+        behavior_ids: Set[str] = set()
+        # TODO nested set comprehension?
+        for step_info in step_infos:
+            behavior_ids |= set(step_info.name_behavior_ids)
```

Comment:
Couldn't we use this to keep `training_behaviors` up to date (cached), and avoid having the callback?

Reply:
That was the approach I wanted to take originally. But at least at the moment, we have to create the trainers (in the trainer_controller) before we get to the StepInfo:
https://github.com/Unity-Technologies/ml-agents/blob/b0be09a0e7cc55a292d63642cfd596e4d2da9c83/ml-agents/mlagents/trainers/trainer_controller.py#L176-L183

Reply:
I think moving:
```python
     external_brain_behavior_ids = set(env_manager.training_behaviors.keys()) 
     new_behavior_ids = external_brain_behavior_ids - last_brain_behavior_ids 
     self._create_trainers_and_managers(env_manager, new_behavior_ids) 
     last_brain_behavior_ids = external_brain_behavior_ids 
```

into `TrainerController.advance` after the `env.advance()` call would solve the ordering issue?

Reply:
(not against changing the abstraction either, just commenting because I put a little thought into this previously)

Reply:
Yeah, I considered that too. It would mean the first step that you get a new behavior, you won't have a trainer to process them, so you'd end up dropping the StepInfo

https://github.com/Unity-Technologies/ml-agents/blob/6dc68df348380152197bc4add9b6073b35d79145/ml-agents/mlagents/trainers/env_manager.py#L114-L120

(we can silence the warning, but still lose some information).

It's not the end of the world, but the callback seems like the lesser of two evils.

Reply:
Ah yeah, I remember this issue.  Seems like the alternative is that the EnvManager has the responsibility for creating the AgentManager, but at the moment the AgentManager depends on the Trainer.